src/zkp/piaffg.rs

```diff
@@ -596,6 +603,353 @@ mod tests {
+        let pk1 = decryption_key_1.encryption_key();
+
+        let X = CurvePoint::GENERATOR.multiply_by_scalar(&x)?;
+        let (Y, rho_y) = pk1
+            .encrypt(&mut rng, &y)
+            .map_err(|_| InternalError::InternalInvariantFailed)?;
+
+        let C = pk0.random_ciphertext(&mut rng);
+
+        let (D, rho) = {
+            let (D_intermediate, rho) = pk0.encrypt(&mut rng, &y).unwrap();
+            let D = pk0.multiply_and_add(&x, &C, &D_intermediate).unwrap();
+            (D, rho)
+        };
+
+        let setup_params = VerifiedRingPedersen::gen(&mut rng, &())?;
+        let input = PiAffgInput::new(setup_params, pk0, pk1, C, D, Y, X);
+        let secret = PiAffgSecret::new(x.clone(), y.clone(), rho.clone(), rho_y.clone());
+
+        // Generate some random elements to use as replacements
+        let random_bignumber = random_plusminus(&mut rng, &k256_order());
+        let random_nonce = Nonce::random(&mut rng, input.prover_encryption_key.modulus());
+
+        // Swap multi coefficient with a random [`BigNumber`]
+        assert_ne!(secret.mult_coeff, random_bignumber);
+        let bad_secret = PiAffgSecret::new(
+            random_bignumber.clone(),
+            y.clone(),
+            rho.clone(),
+            rho_y.clone(),
+        );
+        let bad_proof = PiAffgProof::prove(&input, &bad_secret, &(), &mut transcript(), &mut rng)?;
+        assert!(bad_proof.verify(&input, &(), &mut transcript()).is_err());
+
+        // Swap add coefficient with a random [`BigNumber`]
+        assert_ne!(secret.add_coeff, random_bignumber);
+        let bad_secret = PiAffgSecret::new(x.clone(), random_bignumber, rho.clone(), rho_y.clone());
+        let bad_proof = PiAffgProof::prove(&input, &bad_secret, &(), &mut transcript(), &mut rng)?;
+        assert!(bad_proof.verify(&input, &(), &mut transcript()).is_err());
+
+        // Swap add coefficient nonce verifier key with a random [`Nonce`]
+        assert_ne!(secret.add_coeff_nonce_verifier_key, random_nonce);
+        let bad_secret = PiAffgSecret::new(x.clone(), y.clone(), random_nonce.clone(), rho_y);
+        let bad_proof = PiAffgProof::prove(&input, &bad_secret, &(), &mut transcript(), &mut rng)?;
+        assert!(bad_proof.verify(&input, &(), &mut transcript()).is_err());
+
+        // Swap add coefficient nonce prover key with a random [`Nonce`]
+        assert_ne!(secret.add_coeff_nonce_prover_key, random_nonce);
+        let bad_secret = PiAffgSecret::new(x, y, rho, random_nonce);
+        let bad_proof = PiAffgProof::prove(&input, &bad_secret, &(), &mut transcript(), &mut rng)?;
+        assert!(bad_proof.verify(&input, &(), &mut transcript()).is_err());
+
+        Ok(())
+    }
+
+    #[test]
+    fn negative_test_random_proof_elements() -> Result<()> {
+        let mut rng = init_testing();
+        let x = random_plusminus_by_size(&mut rng, ELL);
+        let y = random_plusminus_by_size(&mut rng, ELL);
```

Comment:
todo: fix `y` range. This will technically always be valid because `ELL < ELL_PRIME` so you're generating an in-range value. But I think it's worth fixing just to avoid confusion in the future about what the expected range is. Plus, having the correct range will ensure that (over time) we randomly test elements across the whole thing including the edge cases.